security/management/commands/purge_logs.py

```diff
@@ -0,0 +1,144 @@
+
+
+storage = import_string(settings.BACKUP_STORAGE_CLASS)()
+
+
+UNIT_OPTIONS = {
+    'h': lambda amount: now() - timedelta(hours=amount),
+    'd': lambda amount: now() - timedelta(days=amount),
+    'w': lambda amount: now() - timedelta(weeks=amount),
+    'm': lambda amount: now() - timedelta(days=(30 * amount)),  # 30-day month
+    'y': lambda amount: now() - timedelta(weeks=(52 * amount)),  # 364-day year
+}
+
+
+def lazy_serialize_qs_without_pk(qs):
+    def generator():
+        for obj in qs.iterator():
+            data = serializers.serialize('python', [obj])[0]
+
+            del data['pk']
+            yield data
+
+    class StreamList(list):
+        def __iter__(self):
+            return generator()
+
+        # according to the comment below
+        def __len__(self):
+            return 1
+
+    return StreamList()
+
+
+class Command(BaseCommand):
+
+    models = {
+        'input-request': InputLoggedRequest,
+        'output-request': OutputLoggedRequest,
+        'command': CommandLog,
+        'celery': CeleryTaskLog
+    }
+
+    def add_arguments(self, parser):
+        super().add_arguments(parser)
+        parser.add_argument('--type', action='store', dest='type', default='input',
+                            help='Tells Django what type of requests should be removed '
+                                 '(input-request/output-request/command/celery).')
+        parser.add_argument('--expiration', action='store', dest='expiration',
+                            help='Sets the timedelta from which logs will be removed.', required=True)
+        parser.add_argument('--noinput', action='store_false', dest='interactive', default=True,
+                            help='Tells Django to NOT prompt the user for input of any kind.')
+        parser.add_argument('--backup', action='store', dest='backup', default=False,
+                            help='Tells Django where to backup removed logs.')
+
+    def backup_logs_and_clean_data(self, qs, backup_path):
+        for timestamp in qs.datetimes('created_at', 'day', tzinfo=utc):
+            min_timestamp = datetime.combine(timestamp, time.min).replace(tzinfo=utc)
+            max_timestamp = datetime.combine(timestamp, time.max).replace(tzinfo=utc)
+            qs_filtered_by_day = qs.filter(
+                **{'created_at__range': (min_timestamp, max_timestamp)}
+            )
+            if backup_path:
+                log_file_name = os.path.join(backup_path, str(timestamp.date()))
+
+                if storage.exists('{}.json.gz'.format(log_file_name)):
+                    i = 1
+                    while storage.exists('{}({}).json.gz'.format(log_file_name, i)):
+                        i += 1
+                    log_file_name = '{}({})'.format(log_file_name, i)
+
+                self.stdout.write(4 * ' ' + 'generate backup file: {}.json.gz'.format(log_file_name))
+
+                with storage.open('{}.json.gz'.format(log_file_name), 'wb') as f:
```

Comment:
it is tested on S3 and it works (everything is stream)

Reply:
It is stream by default?

Reply:
yes

Reply:
file is stream by default